src/ldc/eh/win32.d

```diff
@@ -132,9 +141,89 @@ import rt.util.container.hashtab;
 import core.sync.mutex;
 
 __gshared HashTab!(TypeInfo_Class, _ThrowInfo) throwInfoHashtab;
-__gshared HashTab!(TypeInfo_Class, CatchableType) catchableHashtab;
+__gshared HashTab!(TypeInfo_Class, ImgPtr!(CatchableType*)) catchableHashtab;
 __gshared Mutex throwInfoMutex;
 
+void* safeMalloc(size_t size)
+{
+    if (auto ptr = malloc(size))
+        return ptr;
+    onOutOfMemoryError();
+    return null;
+}
+
+version(Win32)
+{
+    ImgPtr!(T*) eh_malloc(T)(size_t size = T.sizeof)
+    {
+        return cast(T*) safeMalloc(size);
+    }
+
+    T* toPointer(T)(ImgPtr!(T*) imgPtr)
+    {
+        return imgPtr;
+    }
+}
+else
+{
+    /**
+     * Heap dedicated for CatchableTypeArray/CatchableType/TypeDescriptor
+     * structs of cached _ThrowInfos.
+     * The heap is used to keep these structs tightly together, as they are
+     * referenced via 32-bit offsets from a common base. We simply use the
+     * heap's start as base (instead of the actual image base), and malloc()
+     * returns an offset.
+     * The allocated structs are all cached and never released, so this heap
+     * can only grow. The offsets remain constant after a grow, so it's only
+     * the base which may change.
+     */
+    struct EHHeap
+    {
+        void* base;
+        size_t capacity;
+        size_t length;
+
+        this(size_t initialCapacity)
+        {
+            base = safeMalloc(initialCapacity);
+            capacity = initialCapacity;
+        }
+
+        size_t malloc(size_t size)
+        {
+            auto offset = length;
+            enum alignmentMask = size_t.sizeof - 1;
+            auto newLength = (length + size + alignmentMask) & ~alignmentMask;
+            auto newCapacity = capacity;
+            while (newLength > newCapacity)
+                newCapacity *= 2;
+            if (newCapacity != capacity)
+            {
+                auto newBase = safeMalloc(newCapacity);
+                newBase[0 .. length] = base[0 .. length];
+                free(base);
+                base = newBase;
+                capacity = newCapacity;
+            }
+            length = newLength;
+            return offset;
+        }
+    }
+
+    __gshared EHHeap ehHeap;
+
+    ImgPtr!(T*) eh_malloc(T)(size_t size = T.sizeof)
+    {
+        return cast(uint) ehHeap.malloc(size);
+    }
+
+    // NB: The returned pointer may be invalidated by a consequent grow of ehHeap!

```

Comment:
Pointer invalidation ccan also happen for exceptions in flight (in another thread or as a collateral exception).
Even if it's unlikely to hit the reallocation at all with the initial size of 64 kB, maybe we should use GC'd memory instead. 
I was also considering this for the exception stack, as an exception in flight might get collected otherwise. Raising an error from inside the GC might be problematic, though, but these can be preallocated in msvc_eh_init.